src/engraving/rendering/score/horizontalspacing.cpp

```diff
@@ -32,17 +32,767 @@
+{
+    if (springs.empty() || muse::RealIsEqualOrLess(width, 0.0)) {
+        return;
+    }
+
+    std::sort(springs.begin(), springs.end(), [](const Spring& a, const Spring& b) { return a.preTension < b.preTension; });
+    double inverseSpringConst = 0.0;
+    double force = 0.0;
+
+    //! NOTE springs.cbegin() != springs.cend() because of the emptiness check at the top
+    auto spring = springs.cbegin();
+    do {
+        inverseSpringConst += 1 / spring->springConst;
+        width += spring->width;
+        force = width / inverseSpringConst;
+        ++spring;
+    } while (spring != springs.cend() && !(force < spring->preTension));
+
+    for (const Spring& spring2 : springs) {
+        if (force > spring2.preTension) {
+            double newWidth = force / spring2.springConst;
+            spring2.segment->setWidth(newWidth + spring2.segment->widthOffset());
+        }
+    }
+}
+
+void HorizontalSpacing::setPositionsAndWidths(const std::vector<SegmentPosition>& segmentPositions)
+{
+    size_t segmentsSize = segmentPositions.size();
+    for (size_t i = 0; i < segmentsSize; ++i) {
+        Segment* curSeg = segmentPositions[i].segment;
+        double curX = segmentPositions[i].xPosInSystemCoords;
+
+        if (ignoreSegmentForSpacing(curSeg)) {
+            curSeg->setWidth(0.0);
+            continue;
+        }
+
+        Segment* nextSeg = curSeg;
+        double nextX = curX;
+        for (size_t j = i + 1; j < segmentsSize; ++j) {
+            if (!ignoreSegmentForSpacing(segmentPositions[j].segment)) {
+                nextSeg = segmentPositions[j].segment;
+                nextX = segmentPositions[j].xPosInSystemCoords;
+                break;
+            }
+        }
+
+        Measure* curSegMeasure = curSeg->measure();
+        Measure* nextSegMeasure = nextSeg->measure();
+
+        double segWidth = curSegMeasure == nextSegMeasure || nextSeg->isStartRepeatBarLineType() ? nextX - curX : curSeg->minRight();
+
+        curSeg->setWidth(segWidth);
+
+        if (curSegMeasure != nextSegMeasure) {
+            curSegMeasure->setWidth(curX + segWidth - curSegMeasure->x());
+            nextSegMeasure->mutldata()->setPosX(curX + segWidth);
+        }
+
+        curSeg->setXPosInSystemCoords(curX);
+
+        if (nextSeg == curSeg) {
+            double nextSegWidth = nextSeg->minRight();
+            nextSeg->setWidth(nextSegWidth);
+            nextSegMeasure->setWidth(nextX + nextSegWidth - nextSegMeasure->x());
+            nextSeg->setXPosInSystemCoords(nextX);
+        }
+    }
+}
+
+double HorizontalSpacing::getFirstSegmentXPos(Segment* segment, HorizontalSpacingContext& ctx)
+{
+    const MStyle& style = segment->style();
+
+    double x = 0.0;
+    switch (segment->segmentType()) {
+    case SegmentType::BeginBarLine:
+        return 0.0;
+    case SegmentType::ChordRest:
+    {
+        Shape leftBarrier(RectF(0.0, -0.5 * DBL_MAX, 0.0, DBL_MAX));

```

Comment:
> -0.5 * DBL_MAX

Could you explain the purpose of this? :)

Reply:
<img width="245" alt="image" src="https://github.com/user-attachments/assets/82d8ed1f-f5c3-480a-ad57-e75dfc301a32">

It's a bit hacky, but it's also a simple way of ensuring that nothing can cross the left margin of the system (and it was done this way since forever). We basically compute `minLeft` against a "barrier" rectangle of infinite height (but our infinite is actually DBL_MAX, and if we want our rectangle to have height = DBL_MAX we let it start from -0.5 * DBL_MAX so it will extend to both large negative and large positive y).